org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/InvisibleProjectImporter.java

```diff
@@ -91,20 +91,141 @@ public void importToWorkspace(IProgressMonitor monitor) throws OperationCanceled
+			return false;
+		}
+
+		String invisibleProjectName = ProjectUtils.getWorkspaceInvisibleProjectName(rootPath);
+		IProject invisibleProject = ResourcesPlugin.getWorkspace().getRoot().getProject(invisibleProjectName);
+
+		if (!invisibleProject.exists()) {
+			try {
+				JavaLanguageServerPlugin.logInfo("Try to create an invisible project for the workspace " + rootPath);
+				invisibleProject = ProjectUtils.createInvisibleProjectIfNotExist(rootPath);
+				forceUpdateLibPath = true;
+				JavaLanguageServerPlugin.logInfo("Successfully created a workspace invisible project " + invisibleProjectName);
+			} catch (CoreException e) {
+				JavaLanguageServerPlugin.logException("Failed to create the invisible project.", e);
+				return false;
+			}
+		}
+
+		IJavaProject javaProject = JavaCore.create(invisibleProject);
+
+		IFolder workspaceLinkFolder = invisibleProject.getFolder(ProjectUtils.WORKSPACE_LINK);
+		PreferenceManager preferencesManager = JavaLanguageServerPlugin.getPreferencesManager();
+		List<String> sourcePathsFromPreferences = null;
+		if (preferencesManager != null && preferencesManager.getPreferences() != null) {
+			sourcePathsFromPreferences = preferencesManager.getPreferences().getInvisibleProjectSourcePaths();
+		}
+		List<IPath> sourcePaths;
+		if (sourcePathsFromPreferences != null) {
+			sourcePaths = getSourcePaths(sourcePathsFromPreferences, workspaceLinkFolder);
+		} else {
+			IPath relativeSourcePath = sourceDirectory.makeRelativeTo(rootPath);
+			IPath sourcePath = workspaceLinkFolder.getFolder(relativeSourcePath).getFullPath();
+			sourcePaths = Arrays.asList(sourcePath);
+		}
+
+		List<IPath> excludingPaths = getExcludingPath(javaProject, rootPath, workspaceLinkFolder);
+
+		IPath outputPath = getOutputPath(javaProject, preferencesManager.getPreferences().getInvisibleProjectOutputPath(), false /*isUpdate*/);
+
+		IClasspathEntry[] classpathEntries = resolveClassPathEntries(javaProject, sourcePaths, excludingPaths, outputPath);
+		javaProject.setRawClasspath(classpathEntries, outputPath, monitor);
+
+		if (forceUpdateLibPath && preferencesManager != null && preferencesManager.getPreferences() != null) {
+			UpdateClasspathJob.getInstance().updateClasspath(javaProject, preferencesManager.getPreferences().getReferencedLibraries());
+		}
+
+		return true;
+	}
+
+
+	public static IClasspathEntry[] resolveClassPathEntries(IJavaProject javaProject, List<IPath> sourcePaths, List<IPath> excludingPaths, IPath outputPath) throws CoreException {
+		List<IClasspathEntry> newEntries = new LinkedList<>();
+		for (IClasspathEntry entry : javaProject.getRawClasspath()) {
+			if (entry.getEntryKind() != IClasspathEntry.CPE_SOURCE) {
+				newEntries.add(entry);
+			}
+		}
+
+		// Sort the source paths to make the parent folders come first
+		Collections.sort(sourcePaths, new Comparator<IPath>() {
+			@Override
+			public int compare(IPath path1, IPath path2) {
+				return path1.toString().compareTo(path2.toString());
+			}
+		});
```

Comment:
Here is my Java folder structure.
```
.
├── lib
│   ├── commons-lang3-3.10.jar
│   └── gson-2.8.6.jar
└── src
    ├── App.java
    └── main
        └── Hello.java
```

When i add the user setting below, it only adds "src" to source paths, which is not i expected.
```json
    "java.project.sourcePaths": [
        "src/main",
        "src"
    ]
```

You should make the child folders come first and make it have chance to be marked as source path.